cpp/text_module/algorithm/text.cpp

```diff
@@ -32,3 +35,45 @@ void Text::Join(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, m
     return;
   }
 }
+
+void Text::RegexGroups(mgp_list *args, mgp_graph * /*memgraph_graph*/, mgp_result *result, mgp_memory *memory) {
+  mgp::MemoryDispatcherGuard guard{memory};
+  const auto arguments = mgp::List(args);
+  const auto record_factory = mgp::RecordFactory(result);
+
+  try {
+    const std::string input{arguments[0].ValueString()};
+    const std::string regex_str{arguments[1].ValueString()};
+
+    std::regex regex{regex_str};
+
+    size_t mark_count = regex.mark_count();
+    std::vector<int> submatches;
+    for (size_t i = 0; i <= mark_count; ++i) {
+      submatches.emplace_back(i);
+    }
+
+    mgp::List all_results;
+    mgp::List local_results;
+    uint32_t cnt = 1;
+    auto words_begin = std::sregex_token_iterator(input.begin(), input.end(), regex, submatches);
+    auto words_end = std::sregex_token_iterator();
+
+    for (auto it = words_begin; it != words_end; it++, cnt++) {
+      if (auto match = it->str(); !match.empty()) {
+        local_results.AppendExtend(mgp::Value(match));
+      }
+      if (cnt % (mark_count + 1) == 0) {
+        all_results.AppendExtend(mgp::Value(local_results));
+        local_results = mgp::List();
+      }
```

Comment:
I am not sure why do we do this, can you just explain logic here? This should happen on cnt=1, mark_count=0, or cnt=mark_count+1?

Reply:
yes, so the reason for that is Neo4j for every matches expression returns:
[matched_expression, token1, token2, ....m tokenN]
So if mark_count is 2 this means you have 2 tokens in your expression e.g \\w+ and \\d, and hence you want that your list has 3 items. So you will complete your inner list once cnt is 3, 6, 9 and so on.